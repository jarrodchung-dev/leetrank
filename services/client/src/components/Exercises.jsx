import React, { Component } from "react";
import axios from "axios";
import Exercise from "./Exercise.jsx";

class Exercises extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentExercise: 0,
      exercises: [],
      editor: {
        value: "",
        button: { isDisabled: false },
        showGrading: false,
        showCorrect: false,
        showIncorrect: false
      },
      showButtons: { prev: false, next: false }
    };
    this.handleChange = this.handleChange.bind(this);
    this.submitExercise = this.submitExercise.bind(this);
    this.nextExercise = this.nextExercise.bind(this);
    this.prevExercise = this.prevExercise.bind(this);
    this.resetEditor = this.resetEditor.bind(this);
    this.renderButtons = this.renderButtons.bind(this);
  }
  componentDidMount() {
    this.getExercises();
  }
  handleChange(value) {
    const editor = this.state.editor;
    editor.value = value;
    this.setState(editor);
  }
  submitExercise(event, id) {}
  renderButtons() {}
  nextExercise() {}
  prevExercise() {}
  resetEditor() {}
  render() {
    return (
      <div>
        <h1 className="title is-1">Exercises</h1>
        <hr />
        {!this.props.isAuthenticated && (
          <div className="notification is-warning">
            <span>Please log in to submit an exercise.</span>
          </div>
        )}
        {this.state.exercises.length > 0 && (
          <Exercise
            exercise={this.state.exercises[this.state.currentExercise]}
            editor={this.state.editor}
            isAuthenticated={this.props.isAuthenticated}
            onChange={this.onChange}
            submitExercise={this.submitExercise}
          />
        )}
        <div className="field is-grouped">
          {this.state.showButtons.prev && (
            <button className="button is-info" onClick={() => this.prevExercise()}>
              &lt; Prev
            </button>
          )}
          &nbsp; &nbsp;
          {this.state.showButtons.next && (
            <button className="button is-info" onClick={() => this.nextExercise()}>
              Next &gt;
            </button>
          )}
        </div>
      </div>
    );
  }
}

export default Exercises;
