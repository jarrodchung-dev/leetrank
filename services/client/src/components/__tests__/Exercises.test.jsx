import React from "react";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";
import AceEditor from "react-ace";
import Exercises from "../Exercises.jsx";

jest.mock("react-ace");

const exercises = [
  {
    id: 0,
    body: "Define a function called sum"
  },
  {
    id: 1,
    body: "Define a function called reverse"
  },
  {
    id: 2,
    body: "Define a function called factorial"
  }
];

beforeEach(() => {
  console.error = jest.fn();
  console.error.mockClear();
});

describe("<Exercises />", () => {
  it("renders properly when a user is not authenticated", () => {
    const didMount = jest.fn();
    Exercises.prototype.componentDidMount = didMount;
    const wrapper = shallow(<Exercises isAuthenticated={false} />);
    wrapper.setState({ exercises });
    const alert = wrapper.find(".notification");
    expect(alert.length).toBe(1);
    const message = wrapper.find(".notification > span");
    expect(message.get(0).props.children).toContain(
      "Please log in to submit an exercise."
    );
  });

  it("renders properly when a user is authenticated", () => {
    const didMount = jest.fn();
    Exercises.prototype.componentDidMount = didMount;
    const wrapper = shallow(<Exercises isAuthenticated={true} />);
    wrapper.setState({ exercises });
    const alert = wrapper.find(".notification");
    expect(alert.length).toBe(0);
  });

  it("renders a snapshot successfully", () => {
    const didMount = jest.fn();
    Exercises.prototype.componentDidMount = didMount;
    const tree = renderer.create(<Exercises />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("calls componentDidMount when mounted", () => {
    const didMount = jest.fn();
    Exercises.prototype.componentDidMount = didMount;
    const wrapper = mount(<Exercises />);
    expect(didMount).toHaveBeenCalledTimes(1);
  });
});

/*
it("calls componentDidUpdate when mounted", () => {
  const didUpdate = jest.fn();
  Exercises.prototype.componentDidUpdate = didUpdate;
  const wrapper = mount(<Exercises />);
  expect(didUpdate).toHaveBeenCalledTimes(1);
});
*/
