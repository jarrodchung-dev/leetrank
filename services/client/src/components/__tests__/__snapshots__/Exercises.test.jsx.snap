// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Exercises /> renders a snapshot successfully 1`] = `
<div>
  <h1
    className="title is-3"
  >
    Exercises
  </h1>
  <hr />
  <div
    className="notification is-warning"
  >
    <span>
      Please log in to submit an exercise.
    </span>
  </div>
  <div
    className="level"
  >
    <div
      className="level-left"
    >
      <div
        className="level-item"
      />
    </div>
    <div
      className="level-right"
    >
      <div
        className="level-item"
      />
    </div>
  </div>
</div>
`;
